### PR TITLE
Write a test to find sync IO

### PR DESCRIPTION
**Pull request by LukeButters**
# Background

[SC-55504]

Adds a test which finds sync IO.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/AsyncClientAndServiceMustNotUseSyncNetworkIO.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+using NUnit.Framework;
+
+namespace Halibut.Tests
+{
+    public class AsyncClientAndServiceMustNotUseSyncNetworkIO : BaseTest
+    {
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testSyncService: false, testAsyncServicesAsWell: true,
+            testSyncClients: false, testAsyncClients: true,
+            // TODO: ASYNC ME UP!
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/AsyncClientAndServiceMustNotUseSyncNetworkIO.cs:20]: ?

Comment (nathanwoctopusdeploy) [reply source/Halibut.Tests/AsyncClientAndServiceMustNotUseSyncNetworkIO.cs:20]: testAsyncServicesAsWell -> testAsyncServices. Do we need the As Well?

File: source/Halibut.Tests/Support/Streams/NoSyncIOStream.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+using Halibut.Transport.Streams;
+
+namespace Halibut.Tests.Support.Streams
+{
+    public class NoSyncIoStream : DelegateStreamBase
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/Support/Streams/NoSyncIOStream.cs:10]: Can this describe what it actually does? It doesn't do 'NoSyncIo', it just logs when sync IO occurs

File: source/Halibut/HalibutRuntimeBuilder.cs [hunk 3]
```diff
@@ -21,7 +22,19 @@ public class HalibutRuntimeBuilder
         AsyncHalibutFeature asyncHalibutFeature = AsyncHalibutFeature.Disabled;
         Func<string, string, UnauthorizedClientConnectResponse> onUnauthorizedClientConnect;
         HalibutTimeoutsAndLimits halibutTimeoutsAndLimits;
-
+        IBaseStreamWrapper baseStreamWrapper = new NoOpBaseStreamWrapper();
```
Comment (nathanwoctopusdeploy) [inline source/Halibut/HalibutRuntimeBuilder.cs:25]: I'm not too fond of this. Can this not be null if it's not set and when null it does not change the call chains

File: source/Halibut/HalibutRuntimeBuilder.cs [hunk 4]
```diff
@@ -21,7 +22,19 @@ public class HalibutRuntimeBuilder
...
+
+        /// <summary>
+        /// For testing only.
+        /// </summary>
+        /// <param name="baseStreamWrapper"></param>
+        /// <returns></returns>
+        internal HalibutRuntimeBuilder WithBaseStreamWrapper(IBaseStreamWrapper baseStreamWrapper)
```
Comment (nathanwoctopusdeploy) [inline source/Halibut/HalibutRuntimeBuilder.cs:32]: I don't know if BaseStreamWrapper really helps describe what it does. I think a good comment is needed here

**Review by nathanwoctopusdeploy — CHANGES_REQUESTED** [review]
I see the value in this test but I really don't like how invasive it is in Halibut to do this.

Are there other options? Even some nasty reflection to manipulate the Halibut Runtime / manipulate something that is exposed lower down in Halibut. Having to support this from the top down just for a test seems like a lot of noise for future readers to understand the reasoning behind.

**Comment by nathanwoctopusdeploy** [comment]
> I see the value in this test but I really don't like how invasive it is in Halibut to do this.
> 
> Are there other options? Even some nasty reflection to manipulate the Halibut Runtime / manipulate something that is exposed lower down in Halibut. Having to support this from the top down just for a test seems like a lot of noise for future readers to understand the reasoning behind.

Could we do it with a conditional compilation directive? It's invasive and could change behaviour, so we would not want all tests to run like this, but a single test could be run with a specific compile time directive, separate nuke test -> build to compile with the directive etc. When the compilation directive is defined we wrap the base streams #if #endif for all related code and classes so it's guaranteed to be off by default. Constructor could add the IO logging stream to a static collection so the test can access it etc.

This would produce a lot less changes and also mean we only have those changes for one specific test.

File: source/Halibut/Transport/WebSocketConnectionFactory.cs [hunk 5]
```diff
@@ -9,22 +9,25 @@
...
 using Halibut.Transport.Proxy.Exceptions;
+using Halibut.Transport.Streams;
 using Halibut.Util;
 
 namespace Halibut.Transport
 {
     public class WebSocketConnectionFactory : IConnectionFactory
     {
         readonly X509Certificate2 clientCertificate;
+        IStreamFactory streamFactory;
```
Comment (LukeButters) [inline source/Halibut/Transport/WebSocketConnectionFactory.cs:20]: The test caught that we didn't use the factory in all places.

File: source/Halibut/Transport/WebSocketConnectionFactory.cs [hunk 6]
```diff
@@ -55,7 +58,7 @@ public async Task<IConnection> EstablishNewConnectionAsync(ExchangeProtocolBuild
 
             log.Write(EventType.Diagnostic, "Connection established");
 
-            var stream = new WebSocketStream(client);
+            var stream = streamFactory.CreateStream(client);
```
Comment (LukeButters) [inline source/Halibut/Transport/WebSocketConnectionFactory.cs:61]: This is in `EstablishNewConnectionAsync` a new async method.

File: source/Halibut.Tests/AsyncClientAndServiceMustNotUseSyncNetworkIO.cs [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            syncIoRecordingStreamFactory.streams.Count.Should().BeGreaterOrEqualTo(2, "Since we should wrap a stream on the client and the service.");
+        }
+
+        internal static string[] DistinctStackTracesOfSyncIo(SyncIoRecordingStreamFactory syncIoRecordingStreamFactory)
+        {
+            return syncIoRecordingStreamFactory.PlacesSyncIoWasUsed()
+                .Select(n =>
+                {
+                    var lines = n.ToString().Split('\n')
+                        .Where(s => !s.ToString().Contains("System.Threading.Tasks"))
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/AsyncClientAndServiceMustNotUseSyncNetworkIO.cs:76]: Is this just filtering a noisy stack trace?

**Review by nathanwoctopusdeploy — APPROVED** [review]
Nice, LGTM

**Pull request merged**